test/run_test.py

```diff
@@ -47,22 +50,19 @@
 }
 
 
+def print_to_stderr(message):
+    # Print to stderr because test output also goes to stderr. This ensures
+    # synchronization between the two output sources.
+    print(message, file=sys.stderr)
+
+
 def shell(command, cwd):
-    popen = subprocess.Popen(
-        command,
-        stdout=subprocess.PIPE,
-        universal_newlines=True,
-        cwd=cwd,
-        shell=True)
-    for stdout_line in iter(popen.stdout.readline, ''):
-        print(stdout_line.strip('\n'))
-    popen.stdout.close()
-    return_code = popen.wait()
-    return return_code == 0
+    return subprocess.call(
+        shlex.split(command), universal_newlines=True, cwd=cwd) == 0
```

Comment:
I... think this is missing the point of "shell is bad".

It is true that you get to avoid starting a shell process, but honestly, one extra process is not something we actually be worrying about in this context. The *hazard* of shell=True is in the input format, and since you have not changed the input specification of `shell` (it takes a flat command string), the hazard remains.

Let's be a bit more concrete about in what sense shell is bad (and also why, in some sense, it doesn't matter in this use case). And we'll use `shell(cmd, cwd)`as our example.

Suppose that I have some code that looks like `shell("python test.py", ".")`, and I am now asked to extend it to support accepting an extra flag specifying a JUnit XML file to write out; e.g., if I'm writing to foo.xml, I want to run the command `python test.py --junit-xml=foo.xml`. What is the most "obvious" way to extend support for this in the API?

The most obvious way is to use string interpolation to put the parameter in. `"python test.py --junit-xml={}".format(junit_filename)`. This is wrong, if `junit_filename` contains spaces. `"python test.py --junit-xml=\"{}\"".format(junit_filename)` is wrong too, if `junit_filename` contains quotes. No, the only "correct" way to do this in Python 2 is to use `pipes.quote(s) <https://docs.python.org/2/library/pipes.html#pipes.quote>`_ which, for Python 2 compatibility reasons, is in a totally random module and wasn't even publically documented prior to Python 2.7.

The measure of a good API is "how easy is it to do the right thing", and shell strings fail the test.

But I did say it probably doesn't matter. And that's because you are probably not actually setting `--junit-xml` to arbitrary user input, so no one will probably actually trigger the corner case and everything is fine. Until someone copy pastes the code somewhere else where the constraints are different...